Zend/zend_closures.c

```diff
@@ -381,7 +381,45 @@ static ZEND_COLD zend_function *zend_closure_get_constructor(zend_object *object
 static int zend_closure_compare(zval *o1, zval *o2) /* {{{ */
 {
 	ZEND_COMPARE_OBJECTS_FALLBACK(o1, o2);
-	return Z_OBJ_P(o1) != Z_OBJ_P(o2);
+
+	zend_closure *lhs = (zend_closure*) Z_OBJ_P(o1);
+	zend_closure *rhs = (zend_closure*) Z_OBJ_P(o2);
+
+	if (lhs == rhs) {
+	    return SUCCESS;
+	}
+
+	if (!((lhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE) && (rhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE))) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (Z_TYPE(lhs->this_ptr) == IS_OBJECT && Z_TYPE(rhs->this_ptr) == IS_OBJECT) {
+		if (Z_OBJ(lhs->this_ptr) != Z_OBJ(rhs->this_ptr)) {
+			return ZEND_UNCOMPARABLE;
+		}
+	}
+
+	if (lhs->called_scope != rhs->called_scope) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->func.type != rhs->func.type) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->func.common.scope != rhs->func.common.scope) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (lhs->func.type == ZEND_USER_FUNCTION) {
+		return lhs->func.op_array.opcodes != rhs->func.op_array.opcodes;
```

Comment:
What about something like this?
```
trait T {
    public function foo() {}
}
class X {
    use T { T::foo as bar; }
}
```
I believe `X::foo()` and `X::bar()` will have the same opcodes here, but I think they should still be treated as different methods, right?

Reply:
Right, fixed in abf6e9ad583dd6d6565dc2e8704e778693690058